lib/charms/istio_beacon_k8s/v0/service_mesh.py

```diff
@@ -130,33 +218,97 @@ def update_service_mesh(self):
+                            endpoints=policy.endpoints,
+                        ).model_dump()
                     )
-        mesh_rel_data = {
-            "app_name": self._charm.app.name,
-            "model": self._my_namespace(),
-            "policies": policies,
-        }
-        self._relation.data[self._charm.app]["mesh_data"] = json.dumps(mesh_rel_data)
+        self._relation.data[self._charm.app]["policies"] = json.dumps(mesh_policies)
+
+    def _check_cmr(self, mesh_relation: Relation) -> Optional[Relation]:
+        """Check if the given relation is a cmr. If so return the associated cross_model_mesh relation."""
+        for cmr_rel in self._cmr_relations:
+            if cmr_rel.app.name == mesh_relation.app.name:
+                if "cmr_data" not in cmr_rel.data[cmr_rel.app]:
+                    # Data has not yet been provided
+                    continue
+                return cmr_rel
+        return None
 
     def _my_namespace(self):
         """Return the namespace of the running charm."""
         # This method currently assumes the namespace is the same as the model name. We
         # should consider if there is a better way to do this.
         return self._charm.model.name
+
+    def labels(self):
+        """Labels required for a pod to join the mesh."""
+        if self._relation is None or "labels" not in self._relation.data[self._relation.app]:
+            return {}
+        return json.loads(self._relation.data[self._relation.app]["labels"])
+
+
+class ServiceMeshProvider(Object):
+    """Provide a service mesh to applications."""
+
+    def __init__(
+        self, charm: CharmBase, labels: Dict[str, str], mesh_relation_name: str = "service-mesh"
+    ):
+        """Class used to provide information needed to join the service mesh.
+
+        Args:
+            charm: The charm instantiating this object.
+            mesh_relation_name: The relation name as defined in metadata.yaml or charmcraft.yaml
+                for the relation which uses the service_mesh interface.
+            labels: The labels which related applications need to apply to use the mesh.
+        """
+        super().__init__(charm, mesh_relation_name)
+        self._charm = charm
+        self._relation_name = mesh_relation_name
+        self._labels = labels
+        self.framework.observe(
+            self._charm.on[mesh_relation_name].relation_created, self._relation_created
+        )
+
+    def _relation_created(self, _event):
+        self.update_relations()
+
+    def update_relations(self):
+        """Update all relations with the labels needed to use the mesh."""
+        rel_data = json.dumps(self._labels)
+        for relation in self._charm.model.relations[self._relation_name]:
+            relation.data[self._charm.app]["labels"] = rel_data
+
+    def mesh_info(self) -> List[MeshPolicy]:
+        """Return the relation data used to define authorization policies on the mesh."""
```

Comment:
removing "authorization" since its mesh agnostic.  Exact wording is just a suggestion

```suggestion
        """Return the relation data that defines Policies requested by the related applications."""
```